src/Orleans/Runtime/GrainInterfaceMap.cs

```diff
@@ -127,49 +128,68 @@ internal void AddMap(GrainInterfaceMap map)
+        internal void AddEntry(Type iface, Type grain, PlacementStrategy placement, MultiClusterRegistrationStrategy registrationStrategy, bool primaryImplementation)
         {
             lock (this)
             {
-                GrainInterfaceData grainInterfaceData;
+                var grainTypeInfo = grain.GetTypeInfo();
+                var grainName = TypeUtils.GetFullName(grainTypeInfo);
+                var isGenericGrainClass = grainTypeInfo.ContainsGenericParameters;
+                var grainTypeCode = GrainInterfaceUtils.GetGrainClassTypeCode(grain);
 
-                if (table.ContainsKey(interfaceId))
-                {
-                    grainInterfaceData = table[interfaceId];
-                }
-                else
-                {
-                    grainInterfaceData = new GrainInterfaceData(interfaceId, iface, grainInterface);
+                var grainInterfaceData = AddOrGetGrainInterfaceData(iface, isGenericGrainClass);
 
-                    table[interfaceId] = grainInterfaceData;
-                    var interfaceTypeKey = GetTypeKey(iface, isGenericGrainClass);
-                    typeToInterfaceData[interfaceTypeKey] = grainInterfaceData;
-                }
-
-                var implementation = new GrainClassData(grainTypeCode, grainClass, isGenericGrainClass, grainInterfaceData, placement, registrationStrategy);
+                var implementation = new GrainClassData(grainTypeCode, grainName, isGenericGrainClass, grainInterfaceData, placement, registrationStrategy);
                 if (!implementationIndex.ContainsKey(grainTypeCode))
                     implementationIndex.Add(grainTypeCode, implementation);
 
                 grainInterfaceData.AddImplementation(implementation, primaryImplementation);
                 if (primaryImplementation)
                 {
-                    primaryImplementations[grainInterface] = grainClass;
+                    primaryImplementations[grainInterfaceData.GrainInterface] = grainName;
                 }
                 else
                 {
-                    if (!primaryImplementations.ContainsKey(grainInterface))
-                        primaryImplementations.Add(grainInterface, grainClass);
+                    if (!primaryImplementations.ContainsKey(grainInterfaceData.GrainInterface))
+                        primaryImplementations.Add(grainInterfaceData.GrainInterface, grainName);
                 }
 
                 if (localTestMode)
                 {
+                    var assembly = grainTypeInfo.Assembly.CodeBase;
                     if (!loadedGrainAsemblies.Contains(assembly))
                         loadedGrainAsemblies.Add(assembly);
                 }
             }
         }
 
+        private GrainInterfaceData AddOrGetGrainInterfaceData(Type iface, bool isGenericGrainClass)
```

Comment:
Nit: better to name this method GetOrAddXXX, since it only adds if it cannot get